internal/common/customplanmodifier/request_only_required_create.go

```
@@ -0,0 +1,136 @@
+	validateRequestOnlyRequiredOnCreate(
+		isCreate(&req.State),
+		req.PlanValue,
+		req.Path,
+		&resp.Diagnostics,
+	)
+}
+
+func (m *requestOnlyRequiredOnCreateAttributePlanModifier) PlanModifyNumber(ctx context.Context, req planmodifier.NumberRequest, resp *planmodifier.NumberResponse) {
+	validateRequestOnlyRequiredOnCreate(
+		isCreate(&req.State),
+		req.PlanValue,
+		req.Path,
+		&resp.Diagnostics,
+	)
+}
+
+func (m *requestOnlyRequiredOnCreateAttributePlanModifier) PlanModifyList(ctx context.Context, req planmodifier.ListRequest, resp *planmodifier.ListResponse) {
+	validateRequestOnlyRequiredOnCreate(
+		isCreate(&req.State),
+		req.PlanValue,
+		req.Path,
+		&resp.Diagnostics,
+	)
+}
+
+func (m *requestOnlyRequiredOnCreateAttributePlanModifier) PlanModifyMap(ctx context.Context, req planmodifier.MapRequest, resp *planmodifier.MapResponse) {
+	validateRequestOnlyRequiredOnCreate(
+		isCreate(&req.State),
+		req.PlanValue,
+		req.Path,
+		&resp.Diagnostics,
+	)
+}
+
+func (m *requestOnlyRequiredOnCreateAttributePlanModifier) PlanModifySet(ctx context.Context, req planmodifier.SetRequest, resp *planmodifier.SetResponse) {
+	validateRequestOnlyRequiredOnCreate(
+		isCreate(&req.State),
+		req.PlanValue,
+		req.Path,
+		&resp.Diagnostics,
+	)
+}
+
+func (m *requestOnlyRequiredOnCreateAttributePlanModifier) PlanModifyObject(ctx context.Context, req planmodifier.ObjectRequest, resp *planmodifier.ObjectResponse) {
+	validateRequestOnlyRequiredOnCreate(
+		isCreate(&req.State),
+		req.PlanValue,
+		req.Path,
+		&resp.Diagnostics,
+	)
+}
+
+// validateRequestOnlyRequiredOnCreate checks that an attribute has a known, non-null
+// value during create and adds an error if it does not.
+func validateRequestOnlyRequiredOnCreate(isCreate bool, planValue attr.Value, attrPath path.Path, diagnostics *diag.Diagnostics) {
```

should we also throw an error if the user tries to update it? Or is the expectation to always use it in conjunction with customplanmodifier.CreateOnly()?

How is it handled during resource reads? I assume the attribute would still stay in the state even though the GET calls don't return it right?

---

If the user tries to update it, `customplanmodifier.CreateOnly()` is responsible for blocking the change and surfacing an error. `RequestOnlyRequiredOnCreate` only enforces that the attribute is present on create, so for attributes that are both required-on-create and immutable we expect to always use it in conjunction with CreateOnly(). However, an attribute could be `CreateOnly`, without `RequestOnlyRequiredOnCreate`.

Regarding your second question: you're correct. On resource reads, the attribute stays in the state, so it retains its last known value.